terraform/provider/resource_teleport_user.go

```diff
@@ -5,187 +5,141 @@ import (
 
 	"github.com/gravitational/teleport-plugins/terraform/tfschema"
 	"github.com/gravitational/trace"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 
-	"github.com/gravitational/teleport/api/client"
 	"github.com/gravitational/teleport/api/types"
 )
 
+const (
+	userNamePath = "metadata.0.name"
+)
+
 // teleport_user resource definition
 func resourceTeleportUser() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceTeleportUserUpsert,
-		Read:   resourceTeleportUserRead,
-		Update: resourceTeleportUserUpsert,
-		Delete: resourceTeleportUserDelete,
-		Exists: resourceTeleportUserExists,
+		CreateContext: resourceUserCreate,
+		ReadContext:   resourceUserRead,
+		UpdateContext: resourceUserUpdate,
+		DeleteContext: resourceUserDelete,
 
 		Schema: tfschema.SchemaUserV2(),
 	}
 }
 
-func resourceTeleportUserUpsert(d *schema.ResourceData, m interface{}) error {
-
-
-	// Teleport's types package provides functions that initialize
-	// empty structs with their default values.
-	//
-	// This is useful to populate a default struct of the required type,
-	// and then unmarshall terraform data into it, overwriting defaults
-	// when users requested to do so.
-	//
-	// We don't restrict users to only set some subset of "allowed" fields,
-	// granting them flexibility to go crazy with terraform config,
-	// but we want to provide sane defaults that Teleport expects.
-	//
-	// So, CRUD functions should do the following:
-	// 1. Init the object with a provided func, i.e. user := types.NewUser(name)
-	// 2. Set Metadata, Spec, and any other terraform-configured fields into the structs
-	// 3. Send the struct to Teleport's Auth server via Client
-	//
-	// The second part in this list can be fully generated by protoc-gen-schema.
-	//
-
-	// To initialize a types.UserV2 we need their name,
-	// so first we unmarshall the minimal required amount of data.
-
-	metaResourceList := d.Get("metadata").([]interface{})
-	metaResourceData := metaResourceList[0].(map[string]interface{})
-	name := metaResourceData["name"].(string)
-
+// resourceUserCreate creates Teleport user
```

Comment:
This file is very similar to the role resource one, so I'd apply all the comments from there to this one as well.

Reply:
👍